### PR TITLE
Introduction to Hooks

### PR DESCRIPTION
**Pull request by kapnobatai137**
This will be a new workout that introduces Hooks in our React content.

1. Why were hooks needed?
- explained the purpose of hooks
- presented previous methods that achieved the same thing as hooks
  - render props
  - higher-order components
- explained why you would use hooks instead of the methods presented above

2. What are hooks?
- presented the three main types of hooks
- explained what hooks are used for
- added example of state hook
- added counterpart of state hook using class components

3. State hooks

4. Effect hooks

5. Rules of hooks

### CODE REVIEW

**Comment by enki-hq** [comment]
Rule insight:questions failed for files: 
  - javascript/react/introduction-to-hooks/what-are-hooks.md

**Comment by enki-hq** [comment]
Rule insight:standards failed for files: 
  - javascript/react/introduction-to-hooks/effect-hooks.md
  - javascript/react/introduction-to-hooks/rules-of-hooks.md
  - javascript/react/introduction-to-hooks/state-hooks.md

**Comment by enki-hq** [comment]
Rule workout:insights failed for files: 
  - javascript/react/introduction-to-hooks/README.md

**Comment by enki-hq** [comment]
Rule workout:insights failed for files: 
  - javascript/react/introduction-to-hooks/README.md

**Comment by enki-hq** [comment]
### Insight previews:
- [ ] [enkidevs/curriculum/javascript/react/introduction-to-hooks/effect-hooks.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/introduction-to-hooks/effect-hooks.md?ref=77acfd7b0f41ca740e25183a071cefa7f1de15f9)
- [ ] [enkidevs/curriculum/javascript/react/introduction-to-hooks/rules-of-hooks.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/introduction-to-hooks/rules-of-hooks.md?ref=77acfd7b0f41ca740e25183a071cefa7f1de15f9)
- [ ] [enkidevs/curriculum/javascript/react/introduction-to-hooks/state-hooks.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/introduction-to-hooks/state-hooks.md?ref=77acfd7b0f41ca740e25183a071cefa7f1de15f9)
- [ ] [enkidevs/curriculum/javascript/react/introduction-to-hooks/what-are-hooks.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/introduction-to-hooks/what-are-hooks.md?ref=77acfd7b0f41ca740e25183a071cefa7f1de15f9)
- [ ] [enkidevs/curriculum/javascript/react/introduction-to-hooks/why-were-hooks-needed.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/introduction-to-hooks/why-were-hooks-needed.md?ref=77acfd7b0f41ca740e25183a071cefa7f1de15f9)

File: javascript/react/introduction-to-hooks/what-are-hooks.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    return (
+      <div>
+        <p>
+          You clicked {this.state.count}
+          {" "}times
+        </p>
+        <button
+          onClick={() =>
+            this.setState({
+              count: this.state.count + 1
```
Comment (mihaiberq) [inline javascript/react/introduction-to-hooks/what-are-hooks.md]: You might want to have
```
onClick={() => this.setState(state => ({count: state.count + 1}))}
```
as per https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

File: javascript/react/introduction-to-hooks/effect-hooks.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+```
+
+Notice anything new? We have used the `useEffect()` hook to set the document's title to a new custom message that includes the `name` state. If the syntax looks a bit different it is because we are using template literals[1] that allow us to embed expressions. You can think of the `useEffect()` hook as a replacement for `componentDidMount`, `componentDidUpdate`, and `componentWillUnmount` combined.
+
+Note that you have to import the `useEffect` hook from the `'react'` library.
+
+There are two types of effect hooks: with cleanup or without cleanup. In this insight we will focus on the effects without cleanup.
+
+When using class components, the side effects are defined inside the `componentDidMount`, `componentDidUpdate` or `componentWillUnmount` methods, because defining them in the `render` method would be too early (we want side effects to happen **after** React updates the DOM). Here is how our example would look like if we used class components:
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/effect-hooks.md]: What do you think of?

```suggestion
When using class components, a common pattern for side effects is to define them in the `componentDidMount` method, update them in the `componentDidUpdate` method, and clean them up in the `componentWillUnmount` method.

Here is how our example would look like if we used class components:
```

Note: I removed one sentence because we probably don't always want to have side effects after render

File: javascript/react/introduction-to-hooks/effect-hooks.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  render() {
+    return (
+      <p>Hello, my name is {this.name}.</p>
+    );
+  }
+}
+```
+
+You can already see the advantages of using hooks by comparing these two examples. With the `useEffect()` hook, you tell React that your component has to do something after every render. The `useEffect()` hook is placed inside the component so that we can access the `name` variable. This is possible because of how JavaScript's scope works, and it removes the need of React-specific APIs that would have to read the hook. By default, the `useEffect()` is ran after the first render and after every subsequent update.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/effect-hooks.md]: I assume sometime later we'll explain how to control when effects run by using their dependencies?

```suggestion
You can already see the advantages of using hooks by comparing these two examples. With the `useEffect()` hook, you tell React that your component has to do something after every render. The `useEffect()` hook is placed inside the component so that we can access the `name` variable. This is possible because of how JavaScript's scope works, and it removes the need of React-specific APIs that would have to read the hook. By default, the `useEffect()` is ran after the first render and after every subsequent update (we'll see more how to control this later on).
```

File: javascript/react/introduction-to-hooks/effect-hooks.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+* All of the methods mentioned.
+* `componentDidUpdate()`
+* `componentDidMount()`
+* `componentWillUnmount()`
+
+---
+## Revision
+
+Complete the following code to change the document's title to "Welcome to the revision question.":
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/effect-hooks.md]:
```suggestion
Complete the following code to change the document's title to `"Welcome to the revision question."`:
```

File: javascript/react/introduction-to-hooks/rules-of-hooks.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
+---
+author: kapnobatai136
+
+levels:
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/rules-of-hooks.md]: FYI `levels` are fully deprecated, there should be no need to use them anymore.

CC @mihaiberq for confirmation

File: javascript/react/introduction-to-hooks/rules-of-hooks.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    );
+  });
+
+  // ...
+}
+```
+
+Now that you know your functional components can contain multiple hook calls, it is time to learn the rules that apply to hooks.
+
+The first rule is **always call hooks at the top level**. To allow React to correctly preserve the state of hooks between multiple `useState()` and `useEffect` calls, you can't call hooks inside loops, conditions, or nested functions. What this rule does is it ensures that the hooks are called in the exact same order each time a component renders.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/rules-of-hooks.md]: Haven't verified this exact insight but my gut tells me that adding some whitespace would help the readability (feel free to reject this though)

```suggestion
The first rule is **always call hooks at the top level**. 

To allow React to correctly preserve the state of hooks between multiple `useState()` and `useEffect` calls, you can't call hooks inside loops, conditions, or nested functions. What this rule does is it ensures that the hooks are called in the exact same order each time a component renders.
```

File: javascript/react/introduction-to-hooks/rules-of-hooks.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+The second rule is **only call hooks from React functions**. Hooks were designed to work when called either inside React functional components or inside a custom hook (concept which will be taught at a later time). Hooks can't be called from regular JavaScript functions.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/rules-of-hooks.md]: Same thing here

```suggestion
The second rule is **only call hooks from React functions**. 

Hooks were designed to work when called either inside React functional components or inside a custom hook (concept which will be taught at a later time). Hooks can't be called from regular JavaScript functions.
```

File: javascript/react/introduction-to-hooks/rules-of-hooks.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+To help developers follow these rules more easily, the React team has created an ESLint plugin called `eslint-plugin-react-hooks` with plans to include it by default in the Create React App toolkit. To add this to your project, use the following command:
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/rules-of-hooks.md]: Pretty sure this is already in CRA but either way let's avoid temporal language to prevent info becoming stale quickly.

Note: I change the language relating to the installation command but that's not as important, feel free to leave as is if you preer

```suggestion
To help developers follow these rules more easily, the React team has created an ESLint plugin called `eslint-plugin-react-hooks`. A quick way to add it to your project is to install it from NPM[1]:
```

(and maybe add a footnote for NPM?)

File: javascript/react/introduction-to-hooks/state-hooks.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      name: "Enki"
+    };
+  }
+)
+```
+
+But what if you wanted to use functional components? In a function component you can't use the `this` keyword, nor can you assign or read `this.state`. This is where hooks come in:
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/state-hooks.md]: Technically, you can use `this` in any JS function, it's an inherent binding.

Maybe just say

```suggestion
But what if you wanted to use functional components? In a function component you can't assign or read `this.state`. This is where hooks come in:
```

File: javascript/react/introduction-to-hooks/state-hooks.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+function Enki() {
+  const [name, setName] = useState("Enki");
+}
+```
+
+Both of the examples above provide the same functionality. Note that you must import the `useState` hook from the `'react'` library.
+
+By calling the `useState()` hook we declare a *state variable*, which in our case is called `name`. This *state variable* can take any name you want, and it represents a way to preserve some values between the function calls. Basically, by calling `useState()` we enable the functionality of `this.state` for functional components.
+
+The only argument we pass to the `useState()` hook is the **initial state**, and if you want to store two different values in `state` you can call the `useState()` hook twice. The state doesn't have to an object, but it can be if you need it.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/state-hooks.md]: For consistency, pick "we" or "you" (pick whatever you usually use or have seen in our content)

File: javascript/react/introduction-to-hooks/what-are-hooks.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  - '[hooks-intro](https://reactjs.org/docs/hooks-intro.html){documentation}'
+
+---
+
+# What are hooks?
+
+---
+## Content
+
+Hooks represent a new way of using state with your functional components. They are the best approach at reusing stateful logic and enable you to use more of React's features without having to use classes.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/what-are-hooks.md]: Saying they are best sounds a bit subjective. How about this?

```suggestion
Hooks represent a new way of using state with your functional components. They are the great for reusing stateful logic and enable you to use more of React's features without having to use classes.
```

File: javascript/react/introduction-to-hooks/what-are-hooks.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+There are three main types of hooks:
+- State Hooks
+- Effect Hooks
+- Custom Hooks
+
+Before hooks, your components would start out simple but with time they would grow into a random grouping of stateful logic and side effects. Your components could perform some data fetching in the `componentDidMount` or `componentDidUpdate` lifecycle methods, but the `componentDidMount` could also contain some event listeners that were totally unrelated. Hooks introduced a way of splitting up components into smaller functions such that each part is related.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/what-are-hooks.md]: This is a good point.

Here's a quick revision:

```suggestion
Before hooks, many initially simple React component would eventually grow to contain many unrelated groupings of stateful logic and side effects. A component could perform some data fetching in the `componentDidMount` or `componentDidUpdate` lifecycle methods, while the `componentDidMount` could also contain some event listeners. Hooks allow us to separate these unrelated parts by encapsulating related logic into functions.
```

File: javascript/react/introduction-to-hooks/what-are-hooks.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+      </button>
+    </div>
+  );
+}
+```
+
+What the `Enki` component does is that every time the button is clicked, the `count` state is incremented by one. Now, if we would've used class components, the code would've looked something like this:
+
+```jsx
+class Enki extends React.Component {
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/what-are-hooks.md]: We might need a better example here (and above) to better support the [point about grouping unrelated logic](https://github.com/enkidevs/curriculum/pull/1545/files#r326823090)

File: javascript/react/introduction-to-hooks/what-are-hooks.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        >
+          Click me
+        </button>
+      </div>
+    );
+  }
+}
+```
+
+Both components have the same functionality, but you can already see how much easier it is to write state hooks compared to using state in a class component.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/what-are-hooks.md]: ...and then here we can summarize how with hooks related stateful logic and side effects are grouped together and separated from unrelated code.

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+---
+
+# Why were hooks needed?
+
+---
+## Content
+
+Hooks represent a new addition to React, and they allow you to use state or other useful features without using class components.
+
+Hooks were created to solve a number of issues that were encountered in the past five years since React was created.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]: Make this sentence a bit better at standing the test of time:

```suggestion
Hooks were created to solve a number of issues that were encountered in the first five years since React was created.
```

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+In React, there is no direct way to "attach" reusable behavior to a component (such as connecting to a store). The patterns that emerged from the need of reusable behavior were either using [render props](https://reactjs.org/docs/render-props.html) or using [higher-order components](https://reactjs.org/docs/higher-order-components.html) (or HOCs).
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]:
```suggestion
In React, there is no built-in way to "attach" reusable behavior to a component (such as connecting to a store). The patterns that emerged to solve this need were [render props](https://reactjs.org/docs/render-props.html) and [higher-order components](https://reactjs.org/docs/higher-order-components.html) (or HOCs).
```

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+A `render prop` is nothing more than a function that a component uses to know what to render.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]: Could make it a bit more succinct:

```suggestion
A `render prop` is a function that a component uses to know what to render.
```

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+```jsx
+<MyComponent
+  render={person => (
+    <p>Hello {person.name}</p>
+  )}
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]: Add a note here that another version of the "render props" pattern is to pass a function as `children`:

```jsx
<MyComponent>
  {person => (
    <p>Hello {person.name}</p>
  )}
</MyComponent>
```

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+/>
+```
+
+A higher-order component is a function that takes a component and returns a new component:
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]:
```suggestion
A higher-order component is a function that takes a component and returns a new component that includes the reusable behavior
```

File: javascript/react/introduction-to-hooks/why-were-hooks-needed.md [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+```js
+const EnhancedComponent = higherOrderComponent(
+  WrappedComponent
+);
+```
+
+Although you can use these patterns to create reusable behavior, most of the time they require component restructuring. This can make your code hard to follow because your components will be surrounded by wrappers, consumers, higher-order components, render props and any other abstractions that you make.
```
Comment (nem035) [inline javascript/react/introduction-to-hooks/why-were-hooks-needed.md]:
```suggestion
Although you can use these patterns to create reusable behavior, most of the time they require component restructuring. This can make your code more verbose and harder to follow because your components will be surrounded by wrappers, consumers, higher-order components, render props and other, potentially noisy, abstractions.
```

File: javascript/react/introduction-to-hooks/state-hooks.md (hunk 10 above)
Comment (kapnobatai137) [reply javascript/react/introduction-to-hooks/state-hooks.md]: agreed, I'll review the workout and make sure I am consistent

File: javascript/react/introduction-to-hooks/rules-of-hooks.md (hunk 5 above)
Comment (mihaiberq) [reply javascript/react/introduction-to-hooks/rules-of-hooks.md]: Yup, you can omit them

Comment (kapnobatai137) [reply javascript/react/introduction-to-hooks/rules-of-hooks.md]: removed all the `levels` from the insights
